config/initializers/content_security_policy.rb

```diff
@@ -4,17 +4,19 @@
 # For further information see the following documentation
 # https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy
 
-# Rails.application.config.content_security_policy do |policy|
-#   policy.default_src :self, :https
-#   policy.font_src    :self, :https, :data
-#   policy.img_src     :self, :https, :data
-#   policy.object_src  :none
-#   policy.script_src  :self, :https
-#   policy.style_src   :self, :https
+Rails.application.config.to_prepare do
+  asset_host = Rails.configuration.three_scale.asset_host.to_s.strip
+  backend_host = Rails.configuration.backend_client[:url].to_s.strip
 
-#   # Specify URI for violation reports
-#   # policy.report_uri "/csp-violation-report-endpoint"
-# end
+  Rails.application.config.content_security_policy do |policy|
+    policy.default_src :self
+    policy.font_src    :self, asset_host
+    policy.img_src     :self, asset_host, :data
+    policy.script_src  :self, asset_host, :unsafe_inline, :unsafe_eval
+    policy.style_src   :self, asset_host, :unsafe_inline
+    policy.connect_src :self, backend_host
```

Comment:
I think we have a problem with this protection, because it will disallow ANY requests made from ActiveDocs, be it APIcast URL, or the API backend directly.

Also, it would break payment gateway integrations, because they communicate with the gateway backend directly from the browser.

And in fact, Stripe for example, loads also styles and fonts and sources, so it doesn't work with this CSP settings:
![image](https://github.com/3scale/porta/assets/1270328/8e0dd93b-f98d-45cd-8bab-b104a5167792)

Reply:
> I think we have a problem with this protection, because it will disallow ANY requests made from ActiveDocs, be it APIcast URL, or the API backend directly.

Do you mean API Docs? I tested that screen and seems to work fine, I added the Backend URL to the list of allowed hosts.

> Also, it would break payment gateway integrations, because they communicate with the gateway backend directly from the browser.
> 

This could be solved by adding exceptions to the particular controllers. Check the docs: https://guides.rubyonrails.org/security.html#content-security-policy-header

Do you think it's worth working in every exception? or should we desist now before spending more time on this?

Reply:
> Do you mean API Docs? I tested that screen and seems to work fine, I added the Backend URL to the list of allowed hosts.

Well, API docs, but not 3scale's API, but the customer's, so in the admin portal:
- those at `/admin/api_docs/services`
- and scoped by service: `/apiconfig/services/{ID}/api_docs`

plus whatever ActiveDocs (aka swagger) that are rendered on the developer portal.

- Do you think it's worth working in every exception? or should we desist now before spending more time on this?

I would try to find a reasonable compromise... I'd say in the developer portal we should allow everything, because customers may want to load their own sources, styles, fonts, images from CDNs or wherever they want (in addition to what we are aware of, like payment gateways).

As for the admin portal, we could try to list what external calls might be done potentially. But for sure we'd need to allow ANY AJAX calls from the active doc pages.

Maybe let's start with the minimum added security (just enough to silence the warning, but that will not break anything for sure)? And we can keep adding small bits then.

I think that's a bit easier than restrict everything, and then figure out what failed and remove bit by bit.

Reply:
OK, I think I'll allow everything for developer portals. And about admin portal, do you know which use cases do we have to support? So far I tested and fixed:

- API Docs for 3scale API, including AJAX calls to backend
- Calls to CDN

I think I should test the wizard, which will probably fail. Do you know any other scenario?

Reply:
The ones I mentioned in the previous comment:

in the admin portal:
- those at `/admin/api_docs/services`
- and scoped by service: `/apiconfig/services/{ID}/api_docs`

Reply:
I made some changes to implement this policy:

- Admin portal:
  - Fonts, images, scritps and styles allowed from `:self` and CDN
  - All AJAX connections allowed
  - Everything else only allowed from `:self`
- Developer portal:
  - Everything allowed

ActiveDocs screens you mentioned work fine in my tests.

Reply:
Maybe implementing using the rigid Rails mechanism and then trying to improve little by little will ultimately be more painful and slow. But I leave to your discretion.